src/splitter.rs

```diff
@@ -50,4 +128,177 @@ mod tests {
         let vocab = splitter.vocabulary("");
         assert!(vocab.is_empty(), "Empty input should produce empty vocabulary");
     }
+
+    #[test]
+    fn test_phrases_minimum_length() {
+        let splitter = Splitter::new();
+        let text = "hello world";
+        let phrases = splitter.phrases_as_strings(text);
+        
+        assert_eq!(phrases, vec![vec!["hello".to_string(), "world".to_string()]]);
+    }
+
+    #[test]
+    fn test_phrases_no_single_words() {
+        let splitter = Splitter::new();
+        let text = "word";
+        let phrases = splitter.phrases_as_strings(text);
+        
+        assert_eq!(phrases, Vec::<Vec<String>>::new());
+    }
+
+    #[test]
+    fn test_phrases_substrings() {
+        let splitter = Splitter::new();
+        let text = "one two three four";
+        let phrases = splitter.phrases_as_strings(text);
+        
+        let expected = vec![
+            vec!["one".to_string(), "two".to_string()],
+            vec!["one".to_string(), "two".to_string(), "three".to_string()],
+            vec!["one".to_string(), "two".to_string(), "three".to_string(), "four".to_string()],
+            vec!["three".to_string(), "four".to_string()],
+            vec!["two".to_string(), "three".to_string()],
+            vec!["two".to_string(), "three".to_string(), "four".to_string()],
+        ];
+        assert_eq!(phrases, expected);
+    }
+
+    #[test]
+    fn test_phrases_sentence_delimiters() {
+        let splitter = Splitter::new();
+        let text = "hello world. foo bar?";
+        let phrases = splitter.phrases_as_strings(text);
+        
+        let expected = vec![
+            vec!["foo".to_string(), "bar".to_string()],
+            vec!["hello".to_string(), "world".to_string()],
+        ];
+        assert_eq!(phrases, expected);
+    }
+
+    #[test]
+    fn test_phrases_punctuation_removal() {
+        let splitter = Splitter::new();
+        let text = "hello, world! it's working.";
+        let phrases = splitter.phrases_as_strings(text);
+        
+        // This creates one sentence since sentences are split by . ? ; \n\n
+        // So "hello, world! it's working" becomes one sentence that generates all substrings
```

Comment:
This is a good callout - ! should be a sentence delimiter too. I missed that in the requirements. Add that as a requirement and fix.